### ISSUE TITLE
SPH visualization does not work with particle filters

### ISSUE THREAD

**Issue by qobilidop** [yt-project/yt#1797]
### Bug report

**Code for reproduction**

```python
import yt

@yt.particle_filter(requires=['radial_velocity'], filtered_type='gas')
def outflow(pfilter, data):
    return data[pfilter.filtered_type, 'radial_velocity'] > 0

ds = yt.load('FIRE_M12i_ref11/snapshot_600.hdf5')
ds.add_particle_filter('outflow')
yt.ProjectionPlot(ds, 'z', ('outflow', 'density'))
```

**Actual outcome**

```
...
~/project/dev/yt/yt/data_objects/data_containers.py in _generate_container_field(self, field)
    441 
    442     def _generate_container_field(self, field):
--> 443         raise NotImplementedError
    444 
    445     def _parameter_iterate(self, seq):

NotImplementedError: 
```

**Version Information**

  * yt version: yt-4.0 (commit [3d794a1](https://github.com/yt-project/yt/commit/3d794a10a9380c5ca237ff296ec299a05c7840c8))

**Comment by ngoldbaum** [comment]
Making this work would probably require making `ds._sph_ptype` a tuple of strings instead of a single string. We could then add new particle types to the list of known sph particle types.

Right now there can only be one SPH particle type so your filtered particle doesn't act like an SPH particle like you're looking for.

**Comment by qobilidop** [comment]
@ngoldbaum Thanks for the clarification. I'm wondering if it would be easier to add some extra logic to distinguish particle filters of SPH particles. That way we don't need to tweak all the codes where `ds._sph_ptype` appears.

**Comment by ngoldbaum** [comment]
From the point of view of the rest of yt, a particle filter is just a particle type so I don’t know how your extra logic would work. I don’t actually think it would be that bad to allow multiple SPH particles in a single dataset. One issue: should filtered SPH particles use self-consistent smoothing lengths? Or should they inherit smoothing lengths from the full dataset? When we add support for “gather” smoothing how would that work for a filtered particle?

**Comment by qobilidop** [comment]
I see. I need to dive deeper into the code to see what's happening under the hood.

For the smoothing lengths, in the long run, I think it makes sense to make the choice configurable. For example, I'm thinking of passing a parameter around with the value of a specific field to use as the smoothing length, or `'automatic'` to recalculate it self-consistently. For now, using the inherited smoothing length would be easier to implement. And it's useful at least for my own use case.

**Comment by ngoldbaum** [comment]
I definitely agree this is a valid use case and should work. I’ll try to take a closer look at this tomorrow to see if there’s a possible straightforward fix.

**Comment by jzuhone** [comment]
Did either of you figure out something here? I would like to take a stab at it if not.

**Comment by ngoldbaum** [comment]
This requires allowing there to be more than one kind of SPH particle, which no one has worked on yet. There are some tricky aspects to that, for example for gather smoothing you'd probably need a kdtree for kind of SPH particle, I think? That's just to say there might be more complexity here than is immediately obvious.

Please feel free to poke around at this though. If all you want is scatter smoothing I think this should be relatively straightforward, I *think* you'd just need to make `_sph_ptype` a tuple (and possibly rename it `_sph_ptypes` at the same time).

**Comment by ngoldbaum** [comment]
John opened #2250

**Comment by jzuhone** [comment]
@qobilidop so I just accidentally stumbled upon something. It turns out that your example above works in the current `yt-4.0` branch (i.e. without my modifications at all) if you change `filtered_type` from `"gas"` to `"PartType0"`. The problem then is that the default units of plots, etc. are in code units instead of whatever the current default unit system is. 

So there's a chance that what I am doing over in #2250 is overkill and we can find a simpler way to make your above example work as-is.

**Comment by jzuhone** [comment]
Though it looks like this only works for non-derived fields (not surprisingly).

**Comment by qobilidop** [comment]
@jzuhone Thanks for your exploration! That is interesting to know. So the situation could be summarized as this: `"gas"` or `"PartType0"` alone are correctly recognized as sph_ptype; `"PartType0"` + particle filter works too; `"gas"` + particle filter doesn't work. Since `"gas"` is an alias of `"PartType0"`, I guess the real issue might be in the integration of alias ptype and particle filter? I imagine somewhere in the code, it's checking whether the ptype is sph_ptype. It does check whether it is an alias or a particle filter of an sph_ptype, but forget to check the situation where both are involved.

**Comment by jzuhone** [comment]
I think that's a fair summary.

**Comment by jzuhone** [comment]
Closed by #2250.
